### PR TITLE
feat(core): Peer verification on TLS connections

### PR DESCRIPTION
**Pull request by the-nic**
Up to now, the peer certificates in TLS connections have not been verified after establishing the connection to the host. This is not very good practice, as it allows for man in the middle attacks even if the connection is being established.

Therefore, extend NGActiveSSLSocket to enable peer verification for OpenSSL and GnuTLS.
Both will reuse the default CA root trust store in order to verify the certificate.

As an addition, also send Server Name indications to the peer which allows the server we are connecting to to choose the right certificate.

Fixes [#5019](https://sogo.nu/bugs/view.php?id=5019)

### CODE REVIEW

**Comment by the-nic** [comment]
@cgx @extrafu Is there interest in merging/reviewing this? If there are changes needed, I can implement them, but IMHO this is quite important if you want to rely on TLS connections.

**Comment by extrafu** [comment]
I'll have a look at it this week. Thanks!

File: sope-core/NGStreams/NGActiveSSLSocket.m [hunk 1]
```diff
@@ -42,8 +44,36 @@ - (BOOL)primaryConnectToAddress:(id<NGSocketAddress>)_address;
...
+                      onHostName: _hostName];
+  if (![sock connectToAddress:_address]) {
+    NSException *e;
+    e = [[sock lastException] retain];
+    [self release];
+    e = [e autorelease];
+    [e raise];
+    return nil;
+  }
+  return sock;
```
Comment (extrafu) [inline sope-core/NGStreams/NGActiveSSLSocket.m:69]: sock should be autoreleased - class initializers always return autoreleased instances

Comment (the-nic) [reply sope-core/NGStreams/NGActiveSSLSocket.m:70]: Nice catch! Fixed

**Pull request merged**